sway/ipc-i3.c

```diff
@@ -0,0 +1,89 @@
+#include <json-c/json.h>
+#include "sway/config.h"
+#include "sway/ipc-json.h"
+#include "sway/ipc-sway.h"
+#include "sway/tree/view.h"
+
+static void ipc_i3_json_describe_view(struct sway_container *c, json_object *object) {
+	const char *app_id = view_get_app_id(c->view);
+	const char *class = app_id;
+	const char *instance = app_id;
+	const char *role = NULL;
+	uint32_t parent_id = 0;
+	uint32_t window_id = 0;
+
+	(void) role;
+	(void) parent_id;
+	(void) window_id;
+#ifdef HAVE_XWAYLAND
+	if (c->view->type == SWAY_VIEW_XWAYLAND) {
+		class = view_get_class(c->view);
+		instance = view_get_instance(c->view);
+		role = view_get_window_role(c->view);
+		parent_id = view_get_x11_parent_id(c->view);
+		window_id = view_get_x11_window_id(c->view);
+	}
+#endif
+
+	json_object_object_add(object, "window", json_object_new_int(window_id));
+
+	json_object *window_props = json_object_new_object();
+
+	if (class) {
+		json_object_object_add(window_props, "class", json_object_new_string(class));
+	}
+	if (instance) {
+		json_object_object_add(window_props, "instance", json_object_new_string(instance));
+	}
+	if (c->title) {
+		json_object_object_add(window_props, "title", json_object_new_string(c->title));
+	}
+	if (role) {
+		json_object_object_add(window_props, "window_role", json_object_new_string(role));
+	}
+	json_object_object_add(window_props, "transient_for",
+			parent_id ? json_object_new_int(parent_id) : NULL);
+
+	json_object_object_add(object, "window_properties", window_props);
+}
+
+static void ipc_i3_json_describe_container(struct sway_node *node, json_object *object) {
+	struct sway_container *c = node->sway_container;
+	ipc_json_describe_container(node->sway_container, object);
+	if (c->view) {
+		ipc_json_describe_view_common(node->sway_container, object);
+		ipc_i3_json_describe_view(node->sway_container, object);
+	}
+}
+
+struct ipc_json_descriptors ipc_json_i3_descriptors = {
+	.descriptors = {
+		[N_ROOT]        = ipc_json_describe_root,
+		[N_WORKSPACE]   = ipc_json_describe_workspace,
+		[N_OUTPUT]      = ipc_json_describe_output,
+		[N_CONTAINER]   = ipc_i3_json_describe_container,
+	},
+};
+
+static json_object *ipc_i3_get_tree(struct ipc_client *client, char *buf) {
+	return ipc_json_describe_node_recursive(&root->node, &ipc_json_i3_descriptors);
+}
+
+static ipc_handler *i3_commands[] = {
+	[IPC_COMMAND]           = ipc_sway_command,
+	[IPC_GET_BAR_CONFIG]    = ipc_sway_get_bar_config,
+	[IPC_GET_BINDING_MODES] = ipc_sway_get_binding_modes,
+	[IPC_GET_CONFIG]        = ipc_sway_get_config,
+	[IPC_GET_MARKS]         = ipc_sway_get_marks,
+	[IPC_GET_OUTPUTS]       = ipc_sway_get_outputs,
+	[IPC_GET_TREE]          = ipc_i3_get_tree,
+	[IPC_GET_VERSION]       = ipc_sway_get_version,
+	[IPC_GET_WORKSPACES]    = ipc_sway_get_workspaces,
+	[IPC_SEND_TICK]         = ipc_sway_send_tick,
+	[IPC_SUBSCRIBE]         = ipc_sway_subscribe,
```

Comment:
Some enum values are large here (e.g. `IPC_GET_SEATS`), so maybe it's more appropriate to have a list of pairs (command type, handler). Fortunately events (which have a very large value) shouldn't appear here.

Reply:
I think it's mostly fine; the array is still way smaller than a page, and pages are cheap on modern systems anyway. If you're worried about having the array explode in size if somebody mistakenly adds a large value, I can make it size-bound to e.g. 128 slots. I'm not sure using an array of pairs is worth enough the trouble for such a small memory saving, when it ultimately makes the implementation slightly more complicated.

Let me know if you still want me to do so.

Reply:
An implementation on top of ```bsearch``` doesn't look very complicated to me. Use a structure as the array element, add a comparator, ```qsort``` it and access with ```bsearch```. That should be POSIX.1-2001 compliant.

Reply:
Oh, I was not arguing that it was going to be complicated, just *slightly more* complicated than necessary (plus you need to call qsort, as you mentioned, while using a 1:1 array makes it without change compile-time valid; although you could order the array of pairs directly in the source, but this is really not ideal).

My point is that once again one page of .rodata for both arrays is not only cheap but also half relevant for memory usage, as it's file-backed memory anyway, and the kernel may decide to page it out if it runs out of memory.

Reply:
On that note, I also realized that I'm missing a `const` modifier on the array and other global structures introduced.

Reply:
I would rather have a switch..case here as well.